### PR TITLE
SM-781: extracted common Ecrecover func

### PR DESCRIPTION
**Pull request by zakharenkodmytro**
# Proposed Changes
We want to extract the Ecrecover function to shared, as it is already used in several places like benthos-plugin, devices-api and vehicle-signal-decoding.

### CODE REVIEW

File: crypto/crypto.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// VerifySignature godoc
+// checks if the provided signature corresponds to the given message and Ethereum address.
+// payload []byte The original message that was signed.
+// signature string The signature that needs to be verified.
+// ethAddr string The Ethereum address of the signer.
+// @return bool Indicates whether the signature is valid or not.
+// @return error If there was an issue during the verification process.
+func VerifySignature(payload []byte, signature string, ethAddr string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (JamesReate) [reply crypto/crypto.go:48]: I do like using `common.Address` across the board.

File: crypto/crypto.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+const sigLen = 65
+
+var zeroAddr common.Address
+
+// Ecrecover mimics the ecrecover opcode, returning the address that signed
+// hash with signature. sig must have length 65 and the last byte, the recovery
+// byte usually denoted v, must be 27 or 28.
+func Ecrecover(hash, sig []byte) (common.Address, error) {
```
[the remark this answers is not in the document]
Comment (zakharenkodmytro) [reply crypto/crypto.go:17]: This is was an exact copy of https://github.com/DIMO-Network/devices-api/blob/a26b001b1680292ad42ce59e52af7e63970d8cf2/internal/controllers/helpers/handlers.go#L161, but we definitely can improve it

File: crypto/crypto.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zakharenkodmytro) [reply crypto/crypto.go]: updated signature, make sense to me

File: crypto/crypto.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// VerifySignature godoc
+// checks if the provided signature corresponds to the given message and Ethereum address.
+// payload []byte The original message that was signed.
+// signature string The signature that needs to be verified.
+// ethAddr string The Ethereum address of the signer.
+// @return bool Indicates whether the signature is valid or not.
+// @return error If there was an issue during the verification process.
+func VerifySignature(payload []byte, signature string, ethAddr string) (bool, error) {
+	addr := common.HexToAddress(ethAddr)
+	sig := common.FromHex(signature)
```
[the remark this answers is not in the document]
Comment (zakharenkodmytro) [reply crypto/crypto.go]: I think we will delegate validation to the client code

File: crypto/crypto.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	hash := crypto.Keccak256Hash(payload)
```
[the remark this answers is not in the document]
Comment (zakharenkodmytro) [reply crypto/crypto.go]: updated

File: crypto/crypto.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	return crypto.PubkeyToAddress(*pk), nil
+}
+
+// VerifySignature godoc
+// checks if the provided signature corresponds to the given message and Ethereum address.
+// payload []byte The original message that was signed.
+// signature string The signature that needs to be verified.
+// ethAddr string The Ethereum address of the signer.
+// @return bool Indicates whether the signature is valid or not.
+// @return error If there was an issue during the verification process.
```
[the remark this answers is not in the document]
Comment (zakharenkodmytro) [reply crypto/crypto.go]: made description more simple

**Review by JamesReate — APPROVED** [review]
moves refactoring forward. Open to using a Hash type as what Dylan said - on the client side don't know if that makes it easier or harder

**Review by elffjs — APPROVED** [review]

**Pull request merged**